content/reference/java_interop.adoc

```diff
@@ -360,14 +360,28 @@ The best aspect of this is that you need not do anything special in your initial
 100000
 ----
 
+Functions have limited support for primitive arguments and return type: type hints for `long` and `double` (only these) generate primitive-typed overloads. Note that this capability is restricted to functions of arity no greater than 4.
+
+Thus a function defined as
+
+[source,clojure]
+----
+(defn foo ^long [^long n])
+----
+
+both takes and returns values of primitive type `long` (invocations with a boxed argument and indeed any object result in a cast and delegation to the primitive-typed overload).
+
+[NOTE]
+Like function arguments, primitive locals in `loop` bindings are also restricted to the primitive types `long` and `double`.
+
```

Comment:
The NOTE here should be removed - loop bindings can be any primitive type. This is particularly important for tight ^int loops in Java interop, etc.

Reply:
Alex, I thought so too, and was confused when some of my code used
`long` for the loop counter. I then compiled the examples at
https://clojure.org/reference/java_interop#primitives, which all use
`long` as the loop counters as well (though they seem unaware?).

Further experiments and the compiler
https://github.com/clojure/clojure/blob/a19c36927598677c32099dabd0fdb9d3097df259/src/jvm/clojure/lang/Compiler.java#L6377-L6380
convinced me to add the note … I tried but I don’t think a loop binding
can be anything other than `long` and `double`.

Reply:
Of course you know much better than me, so would appreciate having this
cleared up – I will gladly remove the note.

Reply:
@puredanger @glts is actually correct here, see https://dev.clojure.org/jira/browse/CLJ-1905

Reply:
Yeah, I went off on a long tearful bytecode black hole yesterday looking at all that. :) I'm waiting to talk to Rich about it.